### ISSUE TITLE
Restore-DbaDatabase Missing last required TLog backup for point-in-time restore

### ISSUE THREAD

**Issue by alevyinroc** [sqlcollaborative/dbatools#6467]
### Environmental information

```
#### PowerShell version:

Name                           Value                                                                                                                                               
----                           -----                                                                                                                                               
PSVersion                      5.1.14393.3471                                                                                                                                      
PSEdition                      Desktop                                                                                                                                             
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                             
BuildVersion                   10.0.14393.3471                                                                                                                                     
CLRVersion                     4.0.30319.42000                                                                                                                                     
WSManStackVersion              3.0                                                                                                                                                 
PSRemotingProtocolVersion      2.3                                                                                                                                                 
SerializationVersion           1.1.0.1                                                                                                                                             




#### dbatools Module version:


Name    : dbatools
Path    : C:\Users\alevy_admin\Documents\WindowsPowerShell\Modules\dbatools\1.0.104\dbatools.psd1
Version : 1.0.104

Name    : dbatools
Path    : C:\Users\alevy_admin\Documents\WindowsPowerShell\Modules\dbatools\1.0.93\dbatools.psd1
Version : 1.0.93

Name    : dbatools
Path    : C:\Users\alevy_admin\Documents\WindowsPowerShell\Modules\dbatools\1.0.92\dbatools.psd1
Version : 1.0.92

Name    : dbatools
Path    : C:\Users\alevy_admin\Documents\WindowsPowerShell\Modules\dbatools\1.0.91\dbatools.psd1
Version : 1.0.91

Name    : dbatools
Path    : C:\Users\alevy_admin\Documents\WindowsPowerShell\Modules\dbatools\1.0.90\dbatools.psd1
Version : 1.0.90

Name    : dbatools
Path    : C:\Users\alevy_admin\Documents\WindowsPowerShell\Modules\dbatools\1.0.88\dbatools.psd1
Version : 1.0.88

Name    : dbatools
Path    : C:\Users\alevy_admin\Documents\WindowsPowerShell\Modules\dbatools\1.0.83\dbatools.psd1
Version : 1.0.83

Name    : dbatools
Path    : C:\Users\alevy_admin\Documents\WindowsPowerShell\Modules\dbatools\1.0.79\dbatools.psd1
Version : 1.0.79

Name    : dbatools
Path    : C:\Users\alevy_admin\Documents\WindowsPowerShell\Modules\dbatools\1.0.77\dbatools.psd1
Version : 1.0.77

Name    : dbatools
Path    : C:\Users\alevy_admin\Documents\WindowsPowerShell\Modules\dbatools\1.0.75\dbatools.psd1
Version : 1.0.75

Name    : dbatools
Path    : C:\Users\alevy_admin\Documents\WindowsPowerShell\Modules\dbatools\1.0.73\dbatools.psd1
Version : 1.0.73
```
#### SQL Server:

```
Microsoft SQL Server 2017 (RTM-CU18) (KB4527377) - 14.0.3257.3 (X64)   Nov 16 2019 01:14:50   Copyright (C) 2017 Microsoft Corporation  Standard Edition (64-bit) on Windows Server 2016 Standard 10.0 <X64> (Build 14393: ) (Hypervisor) 
```

```sql
/* REPLACE WITH output of @@LANGUAGE */
```

### Report

When attempting to perform a point-in-time restore to a new server using a Maintenance Solution directory structure (no backup history in `msdb` to pull from), the last TLog file is missed in constructing the restore scripts

For example, with TLog backups taken every 15 minutes on the quarter hour, a PIT restore for 19:15 only goes as far as the backup taken at 1945.

#### Host used

- [X] powershell.exe
- [ ] ISE
- [ ] VS Code
- [ ] Other (please specify)

If anything other than powershell.exe was used, please confirm that you can duplicate the issue with powershell.exe

- [ ] Still buggy in powershell.exe

#### Errors Received

None

#### Steps to Reproduce

```
Restore-DbaDatabase -SqlInstance DESTSERVER -DatabaseName MyDBIndexTest -path e:\scratch\backups\ -RestoreTime "2020-04-07 19:55:00" -OutputScriptOnly
```
Results in the following restore script (truncated for readibility)

```
RESTORE DATABASE [MyDBIndexTest] FROM  DISK = N'e:\scratch\backups\Full\SOURCESERVER_MyDBDataProd_FULL_20200406_220404.bak' WITH  FILE = 1,  MOVE N'MyDB_data' TO N'E:\SQL\DATA\MyDBDataProd.mdf',  MOVE N'MyDB_data_log' TO N'E:\SQL\DATA\MyDBDataProd.ldf',  NORECOVERY,  NOUNLOAD,  STATS = 10,  STOPAT = N'04/07/2020 19:55:00'
RESTORE LOG [MyDBIndexTest] FROM  DISK = N'e:\scratch\backups\Log\SOURCESERVER_MyDBDataProd_LOG_20200406_223001.trn' WITH  FILE = 1,  MOVE N'MyDB_data' TO N'E:\SQL\DATA\MyDBDataProd.mdf',  MOVE N'MyDB_data_log' TO N'E:\SQL\DATA\MyDBDataProd.ldf',  NORECOVERY,  NOUNLOAD,  STATS = 10,  STOPAT = N'04/07/2020 

RESTORE LOG [MyDBIndexTest] FROM  DISK = N'e:\scratch\backups\Log\SOURCESERVER_MyDBDataProd_LOG_20200407_193001.trn' WITH  FILE = 1,  MOVE N'MyDB_data' TO N'E:\SQL\DATA\MyDBDataProd.mdf',  MOVE N'MyDB_data_log' TO N'E:\SQL\DATA\MyDBDataProd.ldf',  NORECOVERY,  NOUNLOAD,  STATS = 10,  STOPAT = N'04/07/2020 19:55:00'
RESTORE LOG [MyDBIndexTest] FROM  DISK = N'e:\scratch\backups\Log\SOURCESERVER_MyDBDataProd_LOG_20200407_194500.trn' WITH  FILE = 1,  MOVE N'MyDB_data' TO N'E:\SQL\DATA\MyDBDataProd.mdf',  MOVE N'MyDB_data_log' TO N'E:\SQL\DATA\MyDBDataProd.ldf',  NOUNLOAD,  STATS = 10,  STOPAT = N'04/07/2020 19:55:00'
```

#### Expected Behavior

Expected the TLog from  2020-04-08 20:00 to be restored up to the point of 19:55
#### Actual Behavior

The last TLog used was 2020-04-08 19:45

**Comment by dulrich14** [comment]
After further investigation on this issue I don't think its a problem with Restore-DbaDatabase but with Select-DbaBackupInformation instead. I recreated the issue by doing the following...

Create a database in full mode and create the following backup chain

- 5:29pm Full backup with append (NOINIT)
- 5:30pm Log backup
- 5:35pm Log backup
- 5:40pm Log backup
- 5:41pm Full backup with append (NOINIT)
- 5:45pm Log backup
- 5:50pm Log backup
- 5:54pm Full backup with overwrite and place in different directory (INIT)
- 5:55pm Log backup
- 6:00pm Log backup


Generate several restore scripts with the following command by changing restore times
```powershell
Restore-DbaDatabase -SqlInstance localhost\sqlsb192 -DatabaseName testdb -WithReplace -path 'C:\Path\to\backups\testdb\' -RestoreTime "2020-05-09 18:01:00" -MaintenanceSolutionBackup -OutputScriptOnly
```

Restore script to 5:31pm (Uses oldest backup and needed log backups, looks good)
```sql
RESTORE DATABASE [testdb] FROM  DISK = N'C:\Path\to\backups\testdb\FULL\Instance_testdb_FULL_20200509_172900.bak' WITH  FILE = 1,  MOVE N'testdb' TO N'C:\Path\to\sql\DATA\testdb.mdf',  MOVE N'testdb_log' TO N'C:\Path\to\sql\DATA\testdb_log.ldf',  NORECOVERY,  NOUNLOAD,  REPLACE,  STATS = 10,  STOPAT = N'05/09/2020 17:36:00'
RESTORE LOG [testdb] FROM  DISK = N'C:\Path\to\sql\testdb\LOG\Instance_testdb_LOG_20200509_173500.trn' WITH  FILE = 1,  MOVE N'testdb' TO N'C:\Path\to\sql\DATA\testdb.mdf',  MOVE N'testdb_log' TO N'C:\Path\to\sql\DATA\testdb_log.ldf',  NORECOVERY,  NOUNLOAD,  STATS = 10,  STOPAT = N'05/09/2020 17:36:00'
RESTORE LOG [testdb] FROM  DISK = N'C:\Path\to\backups\testdb\LOG\Instance_testdb_LOG_20200509_174000.trn' WITH  FILE = 1,  MOVE N'testdb' TO N'C:\Path\to\sql\DATA\testdb.mdf',  MOVE N'testdb_log' TO N'C:\Path\to\sql\DATA\testdb_log.ldf',  NOUNLOAD,  STATS = 10,  STOPAT = N'05/09/2020 17:36:00'
```

Restore script to 5:47pm (Uses second full backup and needed log backups, looks good)
```sql
RESTORE DATABASE [testdb] FROM  DISK = N'C:\Path\to\backups\testdb\FULL\Instance_testdb_FULL_20200509_174132.bak' WITH  FILE = 1,  MOVE N'testdb' TO N'C:\Path\to\sql\DATA\testdb.mdf',  MOVE N'testdb_log' TO N'C:\Path\to\sql\DATA\testdb_log.ldf',  NORECOVERY,  NOUNLOAD,  REPLACE,  STATS = 10,  STOPAT = N'05/09/2020 17:47:00'
RESTORE LOG [testdb] FROM  DISK = N'C:\Path\to\backups\testdb\LOG\Instance_testdb_LOG_20200509_174500.trn' WITH  FILE = 1,  MOVE N'testdb' TO N'C:\Path\to\sql\DATA\testdb.mdf',  MOVE N'testdb_log' TO N'C:\Path\to\sql\DATA\testdb_log.ldf',  NORECOVERY,  NOUNLOAD,  STATS = 10,  STOPAT = N'05/09/2020 17:47:00'
RESTORE LOG [testdb] FROM  DISK = N'C:\Path\to\backups\testdb\LOG\Instance_testdb_LOG_20200509_175001.trn' WITH  FILE = 1,  MOVE N'testdb' TO N'C:\Path\to\sql\DATA\testdb.mdf',  MOVE N'testdb_log' TO N'C:\Path\to\sql\DATA\testdb_log.ldf',  NOUNLOAD,  STATS = 10,  STOPAT = N'05/09/2020 17:47:00'
```



Restore script to 6:01pm (Uses second full backup but is missing the last log backup from 6:05pm to roll foward to 6:01pm)
```sql
RESTORE DATABASE [testdb] FROM  DISK = N'C:\Path\to\backups\testdb\FULL\Instance_testdb_FULL_20200509_174132.bak' WITH  FILE = 1,  MOVE N'testdb' TO N'C:\Path\to\sql\MSSQL\DATA\testdb.mdf',  MOVE N'testdb_log' TO N'C:\Path\to\sql\MSSQL\DATA\testdb_log.ldf',  NORECOVERY,  NOUNLOAD,  REPLACE,  STATS = 10,  STOPAT = N'05/09/2020 18:01:00'
RESTORE LOG [testdb] FROM  DISK = N'C:\Path\to\backups\testdb\LOG\Instance_testdb_LOG_20200509_174500.trn' WITH  FILE = 1,  MOVE N'testdb' TO N'C:\Path\to\sql\MSSQL\DATA\testdb.mdf',  MOVE N'testdb_log' TO N'C:\Path\to\sql\MSSQL\DATA\testdb_log.ldf',  NORECOVERY,  NOUNLOAD,  STATS = 10,  STOPAT = N'05/09/2020 18:01:00'
RESTORE LOG [testdb] FROM  DISK = N'C:\Path\to\backups\testdb\LOG\Instance_testdb_LOG_20200509_175001.trn' WITH  FILE = 1,  MOVE N'testdb' TO N'C:\Path\to\sql\MSSQL\DATA\testdb.mdf',  MOVE N'testdb_log' TO N'C:\Path\to\sql\MSSQL\DATA\testdb_log.ldf',  NORECOVERY,  NOUNLOAD,  STATS = 10,  STOPAT = N'05/09/2020 18:01:00'
RESTORE LOG [testdb] FROM  DISK = N'C:\Path\to\backups\testdb\LOG\Instance_testdb_LOG_20200509_175501.trn' WITH  FILE = 1,  MOVE N'testdb' TO N'C:\Path\to\sql\MSSQL\DATA\testdb.mdf',  MOVE N'testdb_log' TO N'C:\Path\to\sql\MSSQL\DATA\testdb_log.ldf',  NORECOVERY,  NOUNLOAD,  STATS = 10,  STOPAT = N'05/09/2020 18:01:00'
RESTORE LOG [testdb] FROM  DISK = N'C:\Path\to\backups\testdb\LOG\Instance_testdb_LOG_20200509_180000.trn' WITH  FILE = 1,  MOVE N'testdb' TO N'C:\Path\to\sql\MSSQL\DATA\testdb.mdf',  MOVE N'testdb_log' TO N'C:\Path\to\sql\MSSQL\DATA\testdb_log.ldf',  NOUNLOAD,  STATS = 10,  STOPAT = N'05/09/2020 18:01:00'
```


Moved missing full backup back into directory. Restore script to 6:01pm (Uses correct full backup and all needed logs)
```sql
RESTORE DATABASE [testdb] FROM  DISK = N'C:\Path\to\backups\testdb\FULL\Instance_testdb_FULL_20200509_175404.bak' WITH  FILE = 1,  MOVE N'testdb' TO N'C:\Path\to\sql\MSSQL\DATA\testdb.mdf',  MOVE N'testdb_log' TO N'C:\Path\to\sql\MSSQL\DATA\testdb_log.ldf',  NORECOVERY,  NOUNLOAD,  REPLACE,  STATS = 10,  STOPAT = N'05/09/2020 18:01:00'
RESTORE LOG [testdb] FROM  DISK = N'C:\Path\to\backups\testdb\LOG\Instance_testdb_LOG_20200509_175501.trn' WITH  FILE = 1,  MOVE N'testdb' TO N'C:\Path\to\sql\MSSQL\DATA\testdb.mdf',  MOVE N'testdb_log' TO N'C:\Path\to\sql\MSSQL\DATA\testdb_log.ldf',  NORECOVERY,  NOUNLOAD,  STATS = 10,  STOPAT = N'05/09/2020 18:01:00'
RESTORE LOG [testdb] FROM  DISK = N'C:\Path\to\backups\testdb\LOG\Instance_testdb_LOG_20200509_180000.trn' WITH  FILE = 1,  MOVE N'testdb' TO N'C:\Path\to\sql\MSSQL\DATA\testdb.mdf',  MOVE N'testdb_log' TO N'C:\Path\to\sql\MSSQL\DATA\testdb_log.ldf',  NORECOVERY,  NOUNLOAD,  STATS = 10,  STOPAT = N'05/09/2020 18:01:00'
RESTORE LOG [testdb] FROM  DISK = N'C:\Path\to\backups\testdb\LOG\Instance_testdb_LOG_20200509_180500.trn' WITH  FILE = 1,  MOVE N'testdb' TO N'C:\Path\to\sql\MSSQL\DATA\testdb.mdf',  MOVE N'testdb_log' TO N'C:\Path\to\sql\MSSQL\DATA\testdb_log.ldf',  NOUNLOAD,  STATS = 10,  STOPAT = N'05/09/2020 18:01:00'
```

The above can also be proven by running the command that feeds the Invoke-DbaAdvancedRestore inside Restore-DBADatabase. Shout out [to this blog post](https://stuart-moore.com/debugging-the-new-dbatools-restore-dbadatabase-pipeline/) for going into detail behind the Restore-DbaDatabase pipeline

```powershell
$BackupHistory = Get-DbaBackupInformation -SqlInstance localhost\sqlsb192  -Path 'C:\Path\to\backups\testdb\'  -MaintenanceSolution
Select-DbaBackupInformation -BackupHistory $BackupHistory -RestoreTime "2020-05-09 18:01:00"
```

The reason why Select-DbaBackupInformation is not grabbing the last log backup needed is due to the Logs Database Backup LSN is not equal to the full backups CheckpointLsn. It is acually greater than in this case because the full backup that overwrote the backup chain is not available. In the table below we can see the Log backup from 6:05pm has a Database Backup LSN of 37000000073900001 > 37000000067300001 (Checkpoint LSN from backup at 5:41).

```powershell
$BackupHistory = Get-DbaBackupInformation -SqlInstance localhost\sqlsb192  -Path 'C:\Path\to\backups\testdb\'  -MaintenanceSolution
$BackupHistory | Select SqlInstance,Database,Type,Start,End, FirstLsn, DatabaseBackupLsn, CheckpointLsn, LastLsn | Format-Table
```

```
SqlInstance  Database Type            Start                End                           FirstLsn DatabaseBackupLsn     CheckpointLsn           LastLsn
-----------  -------- ----            -----                ---                           -------- -----------------     -------------           -------
SQL\Instance testdb   Transaction Log 5/9/2020 6:05:00 PM  5/9/2020 6:05:00 PM  37000000077400001 37000000073900001 37000000077700002 37000000078000001
SQL\Instance testdb   Transaction Log 5/9/2020 6:00:00 PM  5/9/2020 6:00:00 PM  37000000075400001 37000000073900001 37000000075700002 37000000077400001
SQL\Instance testdb   Transaction Log 5/9/2020 5:55:01 PM  5/9/2020 5:55:01 PM  37000000070700001 37000000073900001 37000000073900001 37000000075400001
SQL\Instance testdb   Transaction Log 5/9/2020 5:50:01 PM  5/9/2020 5:50:01 PM  37000000068600001 37000000067300001 37000000068900002 37000000070700001
SQL\Instance testdb   Transaction Log 5/9/2020 5:45:00 PM  5/9/2020 5:45:00 PM  37000000064100001 37000000067300001 37000000067300001 37000000068600001
SQL\Instance testdb   Database        5/9/2020 5:41:32 PM  5/9/2020 5:41:32 PM  37000000067300001 37000000057000001 37000000067300001 37000000067600001
SQL\Instance testdb   Transaction Log 5/9/2020 5:40:00 PM  5/9/2020 5:40:00 PM  37000000060400001 37000000057000001 37000000060700002 37000000064100001
SQL\Instance testdb   Transaction Log 5/9/2020 5:35:00 PM  5/9/2020 5:35:00 PM  37000000052200001 37000000057000001 37000000057000001 37000000060400001
SQL\Instance testdb   Database        5/9/2020 5:29:00 PM  5/9/2020 5:29:00 PM  37000000057000001 37000000022900001 37000000057000001 37000000057300001
```

Long story short, the Select-DBABackupInformation should allow for CheckpointLsn's that are -ge and not just -eq. See pull request [6523](https://github.com/sqlcollaborative/dbatools/pull/6523)

**Comment by wsmelton** [comment]
And repeating your tests with adding a differential backup, does that still provide the required log backups?

**Comment by dulrich14** [comment]
After testing with differential backups, the necessary log backups are still included and work as expected. 

Backup history with full backup removed from folder at 5:56pm

```
SqlInstance  Database Type                  TotalSize DeviceType Start                   Duration End                    
-----------  -------- ----                  --------- ---------- -----                   -------- ---                    
SQL\Instance testdb   Transaction Log       76.00 KB  Disk       2020-05-10 18:05:00.000 00:00:00 2020-05-10 18:05:00.000
SQL\Instance testdb   Database Differential 784.00 KB Disk       2020-05-10 18:01:16.000 00:00:00 2020-05-10 18:01:16.000
SQL\Instance testdb   Transaction Log       76.00 KB  Disk       2020-05-10 18:00:00.000 00:00:00 2020-05-10 18:00:00.000
SQL\Instance testdb   Transaction Log       76.00 KB  Disk       2020-05-10 17:55:00.000 00:00:00 2020-05-10 17:55:00.000
SQL\Instance testdb   Transaction Log       76.00 KB  Disk       2020-05-10 17:50:00.000 00:00:00 2020-05-10 17:50:00.000
SQL\Instance testdb   Database Differential 784.00 KB Disk       2020-05-10 17:45:14.000 00:00:00 2020-05-10 17:45:14.000
SQL\Instance testdb   Database              2.89 MB   Disk       2020-05-10 17:43:31.000 00:00:00 2020-05-10 17:43:31.000
```

Results from the following powershell. Note how the most recent differential is left out since the full backup is missing from 5:56pm.
```Powershell
$BackupHistory = Get-DbaBackupInformation -SqlInstance localhost\sqlsb192  -Path 'C:\path\to\backups\testdb\'  -MaintenanceSolution
Select-DbaBackupInformation -BackupHistory $BackupHistory -RestoreTime "2020-05-10 18:04:00"
```
```
SqlInstance  Database Type                  TotalSize DeviceType Start                   Duration End                    
-----------  -------- ----                  --------- ---------- -----                   -------- ---                    
SQL\Instance testdb   Database              2.89 MB   Disk       2020-05-10 17:43:31.000 00:00:00 2020-05-10 17:43:31.000
SQL\Instance testdb   Database Differential 784.00 KB Disk       2020-05-10 17:45:14.000 00:00:00 2020-05-10 17:45:14.000
SQL\Instance testdb   Transaction Log       76.00 KB  Disk       2020-05-10 17:50:00.000 00:00:00 2020-05-10 17:50:00.000
SQL\Instance testdb   Transaction Log       76.00 KB  Disk       2020-05-10 17:55:00.000 00:00:00 2020-05-10 17:55:00.000
SQL\Instance testdb   Transaction Log       76.00 KB  Disk       2020-05-10 18:00:00.000 00:00:00 2020-05-10 18:00:00.000
```

and when adding back in the missing full database backup, the command uses the shortest chain possible as expected. 

```Powershell
$BackupHistory = Get-DbaBackupInformation -SqlInstance localhost\sqlsb192  -Path 'C:\path\to\backups\testdb\'  -MaintenanceSolution
Select-DbaBackupInformation -BackupHistory $BackupHistory -RestoreTime "2020-05-10 18:04:00"
```

```
SqlInstance  Database Type                  TotalSize DeviceType Start                   Duration End                    
-----------  -------- ----                  --------- ---------- -----                   -------- ---                    
SQL\Instance testdb   Database              2.89 MB   Disk       2020-05-10 17:56:08.000 00:00:00 2020-05-10 17:56:08.000
SQL\Instance testdb   Database Differential 784.00 KB Disk       2020-05-10 18:01:16.000 00:00:00 2020-05-10 18:01:16.000
SQL\Instance testdb   Transaction Log       76.00 KB  Disk       2020-05-10 18:05:00.000 00:00:00 2020-05-10 18:05:00.000
```

I also tested out the actual restores to confirm they worked, not just the output of scripts.
